### PR TITLE
DRA: remove "source" indirection from v1 Pod API

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup
/kind api-change

#### What this PR does / why we need it:

This makes the API nicer:

    resourceClaims:
    - name: with-template
      resourceClaimTemplateName: test-inline-claim-template
    - name: with-claim
      resourceClaimName: test-shared-claim

Previously, this was:

    resourceClaims:
    - name: with-template
      source:
        resourceClaimTemplateName: test-inline-claim-template
    - name: with-claim
      source:
        resourceClaimName: test-shared-claim

A more long-term benefit is that other, future alternatives might not make sense under the "source" umbrella.

#### Special notes for your reviewer:

This is a breaking change. It's justified because DRA is still alpha and will have several other API breaks in 1.31.

#### Does this PR introduce a user-facing change?

```release-note
DRA: in the `pod.spec.recourceClaims` array, the `source` indirection is no longer necessary. Instead of e.g. `source: resourceClaimTemplateName: my-template`, one can write `resourceClaimTemplateName: my-template`.
```

### CODE REVIEW

**Review by thockin — COMMENTED** [review]
LGTM from API perspective

**Comment by thockin** [comment]
Do you also want me to do full code review or @johnbelamaric can you help here?

**Comment by johnbelamaric** [comment]
I can look in a bit but I am not an approver.

**Comment by thockin** [comment]
If you are happy with the logic, I am happy with API, and I can do final
approve on most of it

**Review by johnbelamaric — APPROVED** [review]
LGTM, changes are mostly mechanical, no logic changes.

/approve
/lgtm

File: pkg/apis/core/validation/validation.go [hunk 1]
```diff
@@ -3020,35 +3020,28 @@ func validatePodResourceClaim(podMeta *metav1.ObjectMeta, claim core.PodResource
 		nameErrs := ValidateDNS1123Label(claim.Name, fldPath.Child("name"))
 		if len(nameErrs) > 0 {
 			allErrs = append(allErrs, nameErrs...)
-		} else if podMeta != nil && claim.Source.ResourceClaimTemplateName != nil {
+		} else if podMeta != nil && claim.ResourceClaimTemplateName != nil {
 			claimName := podMeta.Name + "-" + claim.Name
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/apis/core/validation/validation.go:3024]: We already stopped doing this deterministic naming thing several releases ago. I just forgot to remove this particular piece of validation for it :sweat: 

I'll fix this in this PR as I am touching these lines, even though it is a bit unrelated.

[the remark this answers is not in the document]
Comment (pohly) [reply pkg/apis/core/validation/validation.go]: Done, see additional commit.

I removed the validation test that was specifically for the combination of pod name + claim name. There are other test cases for the rest of the validation (uniqueness, bad name, etc.).

**Comment by thockin** [comment]
Thanks!

/lgtm
/approve

**Pull request merged**